dali/operators/decoder/nvjpeg/permute_layout.cu

```diff
@@ -12,40 +12,102 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "dali/operators/decoder/nvjpeg/permute_layout.h"
+#include <cassert>
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
 #include "dali/core/static_switch.h"
 #include "dali/core/util.h"
-#include "dali/core/format.h"
-#include "dali/core/error_handling.h"
-
+#include "dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.cuh"
+#include "dali/operators/decoder/nvjpeg/permute_layout.h"
 
 namespace dali {
 
-template <int64_t C, typename T>
-__global__ void permuteToInterleavedK(T *output, const T *input, int64_t comp_size) {
+template <int C, typename Output, typename Input>
+__global__ void planar_to_interleaved(Output *output, const Input *input, int64_t npixels) {
   auto tid = blockIdx.x * blockDim.x + threadIdx.x;
-  if (tid >= comp_size) return;
-  T *out = output + C * tid;
+  if (tid >= npixels) return;
+  Output *out = output + C * tid;
   for (int c = 0; c < C; ++c) {
-    auto l = input[c * comp_size + tid];
-    out[c] = l;
+    out[c] = input[c * npixels + tid];
   }
 }
 
-void PermuteToInterleaved(uint8_t *output, const uint8_t *input,
-                          int64_t comp_size, int64_t comp_count, cudaStream_t stream) {
+template <typename Output, typename Input>
+__global__ void planar_rgb_to_bgr(Output *output, const Input *input, int64_t npixels) {
+  auto tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid >= npixels) return;
+  Output r = ConvertNorm<Output>(input[tid]);
+  Output g = ConvertNorm<Output>(input[tid + npixels]);
+  Output b = ConvertNorm<Output>(input[tid + 2 * npixels]);
+  Output *out = output + 3 * tid;
+  out[0] = b;
+  out[1] = g;
+  out[2] = r;
+}
+
+template <typename Output, typename Input>
+__global__ void planar_rgb_to_ycbcr(Output *output, const Input *input, int64_t npixels) {
+  auto tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid >= npixels) return;
+  Output r = ConvertNorm<Output>(input[tid]);
+  Output g = ConvertNorm<Output>(input[tid + npixels]);
+  Output b = ConvertNorm<Output>(input[tid + 2 * npixels]);
+  Output *out = output + 3 * tid;
+  out[0] = kernels::rgb_to_y<Output>({r, g, b});
```

Comment:
Do we need to convert before feeding to rgb_to_y, or maybe it should accept any type and only convert the final result?

Reply:
:+1: